porch/controllers/klippy/pkg/controllers/klippy/klippy_controller.go

```diff
@@ -0,0 +1,436 @@
+	}
+
+	isMatch := false
+	bindings := make(map[schema.GroupKind]*unstructured.Unstructured)
+	{
+		for _, obj := range parentObjects {
+			annotations := obj.GetAnnotations()
+			if annotations["config.kubernetes.io/local-config"] == "binding" {
+				// Ignore binding objects in the "parent"; they don't satisfy bindings
+				continue
+			}
+
+			gk := obj.GroupVersionKind().GroupKind()
+
+			requirements, found := slots[gk]
+			if !found {
+				continue
+			}
+
+			matchesLabels := true
+			for k, v := range requirements.MatchLabels {
+				if obj.GetLabels()[k] != v {
+					matchesLabels = false
+				}
+			}
+			if !matchesLabels {
+				continue
+			}
+			bindings[gk] = obj
+		}
+
+		isMatch = len(bindings) == len(slots)
+		if !isMatch {
+			log.Info("package does not match", "parent", parent.Spec.PackageName, "child", blueprint.PackageName)
+		}
+	}
+
+	if isMatch {
+		log.Info("matched package", "parent", parent.Spec.PackageName, "child", blueprint.PackageName)
+		// TODO: How to avoid collisions
+		name := "packagename-" + strings.ReplaceAll(blueprint.PackageName, "/", "-")
+		packageName := blueprint.PackageName
+
+		clone := &api.PackageCloneTaskSpec{}
+		clone.Upstream = api.UpstreamPackage{
+			UpstreamRef: &api.PackageRevisionRef{
+				Name: blueprint.ID.Name,
+			},
+		}
+
+		parentName := parent.Spec.PackageName
+		objectName := packageName + "-" + parentName + "-" + name
+
+		// TODO: How to sanitize?
+		// packageName can be something like dir/name, and that isn't allowed in names
+		objectName = strings.ReplaceAll(objectName, "/", "-")
+
+		proposal := &api.PackageRevision{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "PackageRevision",
+				APIVersion: api.SchemeGroupVersion.Identifier(),
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: parent.GetNamespace(),
+				Name:      objectName,
+			},
+			Spec: api.PackageRevisionSpec{
+				PackageName:    name,
+				Revision:       "v1",
+				RepositoryName: parent.Spec.RepositoryName,
+				// TODO: Speculative: true,
+				Tasks: []api.Task{
+					{
+						Type:  api.TaskTypeClone,
+						Clone: clone,
+					},
+				},
+			},
+		}
+		proposal.Labels = map[string]string{
+			"alpha.kpt.dev/proposal": "",
+		}
+		proposal.Spec.Parent = &api.ParentReference{
+			Name: parent.Name,
+		}
+
+		for _, bindingObj := range bindings {
+			bindingCore := &unstructured.Unstructured{}
+			bindingCore.SetAPIVersion(bindingObj.GetAPIVersion())
+			bindingCore.SetKind(bindingObj.GetKind())
+			bindingCore.SetName(bindingObj.GetName())
+			if bindingObj.GetNamespace() != "" {
+				bindingCore.SetNamespace(bindingObj.GetNamespace())
+			}
+
+			var bindingConfig runtime.RawExtension
+			bindingConfig.Object = bindingCore
+
+			image := "bind"
+
+			proposal.Spec.Tasks = append(proposal.Spec.Tasks, api.Task{
+				Type: api.TaskTypeEval,
+				Eval: &api.FunctionEvalTaskSpec{
+					Image:  image,
+					Config: bindingConfig,
+				},
+			})
+
+		}
+
+		proposals = append(proposals, proposal)
+	}
```

Comment:
lots of binding stuff that I skipped over. I am curious, where is `bind` function used that was taken as input ?

Reply:
Ooops - I forgot to do that - fixed in the latest update!